### PR TITLE
Introduce a long lived section of the heap.

### PR DESCRIPTION
**Pull request by tannewt**
This adapts the allocation process to start from either end of the heap
when searching for free space. The default behavior is identical to the
existing behavior where it starts with the lowest block and looks higher.
Now it can also look from the highest block and lower depending on the
long_lived parameter to gc_alloc. As the heap fills, the two sections may
overlap. When they overlap, a collect may be triggered in order to keep
the long lived section compact. However, free space is always eligable
for each type of allocation.

Heap prior would end up looking something like: 
heap_layout

Afterwards its:
heap_layout1819

Video of it working here: https://www.youtube.com/watch?v=S0uEZqxOWOc

By starting from either of the end of the heap we have ability to separate
short lived objects from long lived ones. This separation reduces heap
fragmentation because long lived objects are easy to densely pack.

Most objects are short lived initially but may be made long lived when
they are referenced by a type or module. This involves copying the
memory and then letting the collect phase free the old portion.

QSTR pools and chunks are always long lived because they are never freed.

The reallocation, collection and free processes are largely unchanged. They
simply also maintain an index to the highest free block as well as the lowest.
These indices are used to speed up the allocation search until the next collect.

In practice, this change may slightly slow down import statements with the
benefit that memory is much less fragmented afterwards. For example, a test
import into a 20k heap that leaves ~6k free previously had the largest
continuous free space of ~400 bytes. After this change, the largest continuous
free space is over 3400 bytes.

### CODE REVIEW

**Comment by tannewt** [comment]
I'm still looking into fixing the tests so sit tight.

**Comment by dhalbert** [comment]
I would like to go first :)

**Comment by tannewt** [comment]
Ok, this is ready for review.

File: py/gc.c [hunk 1]
```diff
@@ -152,14 +146,19 @@ void gc_init(void *start, void *end) {
     memset(MP_STATE_MEM(gc_finaliser_table_start), 0, gc_finaliser_table_byte_len);
 #endif
 
-    // set last free ATB index to start of heap
+    // Set first free ATB index to the start of the heap.
     MP_STATE_MEM(gc_last_free_atb_index) = 0;
+    // Set last free ATB index to the end of the heap.
+    MP_STATE_MEM(gc_last_free_atb_index) = MP_STATE_MEM(gc_alloc_table_byte_len) - 1;
```
Comment (dhalbert) [inline py/gc.c]: line 150 and 152 are both setting `MP_STATE_MEM(gc_last_free_atb_index)`, so line 150 is wrong or redundant?

File: py/gc_long_lived.c [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    for (uint32_t i = 0; i < gc_nbytes(fun_bc->const_table) / sizeof(mp_obj_t); i++) {
+        // Skip things that aren't allocated on the heap (and hence have zero bytes.)
+        if (gc_nbytes((byte *)fun_bc->const_table[i]) == 0) {
+            continue;
+        }
+        // Try to detect raw code.
+        mp_raw_code_t* raw_code = MP_OBJ_TO_PTR(fun_bc->const_table[i]);
+        if (raw_code->kind == MP_CODE_BYTECODE) {
+            raw_code->data.u_byte.bytecode = gc_make_long_lived((byte*) raw_code->data.u_byte.bytecode);
+            // TODO(tannewt): Do we actually want to recurse here?
```
Comment (dhalbert) [inline py/gc_long_lived.c]: Is this still a question?

File: py/gc_long_lived.c [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            raw_code->data.u_byte.const_table = gc_make_long_lived((byte*) raw_code->data.u_byte.const_table);
+        }
+        ((mp_uint_t *) fun_bc->const_table)[i] = (mp_uint_t) make_obj_long_lived(
+            (mp_obj_t) fun_bc->const_table[i], max_depth - 1);
+
+    }
+    fun_bc->const_table = gc_make_long_lived((mp_uint_t*) fun_bc->const_table);
+    // extra_args stores keyword only argument default values.
+    size_t words = gc_nbytes(fun_bc) / sizeof(mp_uint_t*);
+    for (size_t i = 0; i < words - 4; i++) {
```
Comment (dhalbert) [inline py/gc_long_lived.c]: What's the `4`? Is that number of bytes? could it be 8 on 64-bit-word impls?

File: py/gc.c (hunk 1 above)
Comment (tannewt) [reply py/gc.c]: 150 was wrong. Good catch!

File: py/gc_long_lived.c (hunk 2 above)
Comment (tannewt) [reply py/gc_long_lived.c]: I'm still unsure about it but the comment isn't useful so I removed it.

File: py/gc_long_lived.c (hunk 3 above)
Comment (tannewt) [reply py/gc_long_lived.c]: Its the number of pointers stored in mp_obj_fun_bc_t before the extra_args array. Is there another way to get the array length?

Comment (dhalbert) [reply py/gc_long_lived.c]: The struct defn is:
```
typedef struct _mp_obj_fun_bc_t {
    mp_obj_base_t base;
    mp_obj_dict_t *globals;         // the context within which this function was defined
    const byte *bytecode;           // bytecode for the function
    const mp_uint_t *const_table;   // constant table
    // the following extra_args array is allocated space to take (in order):
    //  - values of positional default args (if any)
    //  - a single slot for default kw args dict (if it has them)
    //  - a single slot for var args tuple (if it takes them)
    //  - a single slot for kw args dict (if it takes them)
    mp_obj_t extra_args[];
} mp_obj_fun_bc_t;
```
I'm not sure why it doesn't say `[4]`. then I think you could use `sizeof()`. And if it's a VLA (variable length array), you can use `sizeof()` also. Found this: https://stackoverflow.com/questions/14995870/behavior-of-sizeof-on-variable-length-arrays-c-only.

But I'm not sure this is worth fixing.

Comment (dhalbert) [reply py/gc_long_lived.c]: I guess it's allocated separately and assigned to there?

Comment (tannewt) [reply py/gc_long_lived.c]: Its done through a cast so I'm not sure if sizeof would work: https://github.com/adafruit/circuitpython/blob/master/py/objfun.c#L356

**Pull request merged**